<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to pathfindR • pathfindR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Introduction to pathfindR">
<meta property="og:description" content="pathfindR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathfindR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.6.3.9006</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/comparing_results.html">Comparing Two pathfindR Results</a>
    </li>
    <li>
      <a href="../articles/intro_vignette.html">Introduction to pathfindR</a>
    </li>
    <li>
      <a href="../articles/manual_execution.html">Step-by-Step Execution of the pathfindR Enrichment Workflow</a>
    </li>
    <li>
      <a href="../articles/non_hs_analysis.html">pathfindR Analysis for non-Homo-sapiens organisms</a>
    </li>
    <li>
      <a href="../articles/obtain_data.html">Obtaining PIN and Gene Sets Data</a>
    </li>
    <li>
      <a href="../articles/visualization_vignette.html">Visualization of pathfindR Enrichment Results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/egeulgen/pathfindR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="intro_vignette_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to pathfindR</h1>
                        <h4 data-toc-skip class="author">Ege Ulgen</h4>
            
            <h4 data-toc-skip class="date">2022-06-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/egeulgen/pathfindR/blob/HEAD/vignettes/intro_vignette.Rmd" class="external-link"><code>vignettes/intro_vignette.Rmd</code></a></small>
      <div class="hidden name"><code>intro_vignette.Rmd</code></div>

    </div>

    
    
<p><code>pathfindR</code> is a tool for enrichment analysis via active subnetworks. The package also offers functionalities to cluster the enriched terms and identify representative terms in each cluster, to score the enriched terms per sample and to visualize analysis results.</p>
<p>The functionalities of pathfindR is described in detail in <em>Ulgen E, Ozisik O, Sezerman OU. 2019. pathfindR: An R Package for Comprehensive Identification of Enriched Pathways in Omics Data Through Active Subnetworks. Front. Genet. <a href="https://doi.org/10.3389/fgene.2019.00858" class="external-link">https://doi.org/10.3389/fgene.2019.00858</a></em></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The observation that motivated us to develop <code>pathfindR</code> was that direct enrichment analysis of differential RNA/protein expression or DNA methylation results may not provide the researcher with the full picture. That is to say: enrichment analysis of only a list of significant genes alone may not be informative enough to explain the underlying disease mechanisms. Therefore, we considered leveraging interaction information from a protein-protein interaction network (PIN) to identify distinct active subnetworks and then perform enrichment analyses on these subnetworks.</p>
<blockquote>
<p>An active subnetwork can be defined as a group of interconnected genes in a PIN that predominantly consists of significantly altered genes. In other words, active subnetworks define distinct disease-associated sets of interacting genes, whether discovered through the original analysis or discovered because of being in interaction with a significant gene.</p>
</blockquote>
<p>The active-subnetwork-oriented enrichment analysis approach of pathfindR can be summarized as follows: Mapping the input genes with the associated p values onto the PIN (after processing the input), active subnetwork search is performed. The resulting active subnetworks are then filtered based on their scores and the number of significant genes they contain. This filtered list of active subnetworks are then used for enrichment analyses, i.e. using the genes in each of the active subnetworks, the significantly enriched terms (pathways/gene sets) are identified. Enriched terms with adjusted p values larger than the given threshold are discarded and the lowest adjusted p value (over all active subnetworks) for each term is kept. This process of <code>active subnetwork search + enrichment analyses</code> is repeated for a selected number of iterations, performed in parallel. Over all iterations, the lowest and the highest adjusted-p values, as well as number of occurrences over all iterations are reported for each significantly enriched term in the resulting data frame. An HTML report containing the results is also provided containing links to the visualizations of the enriched terms. This active-subnetwork-oriented enrichment approach is demonstrated in the section <a href="#active-subnetwork-oriented-enrichment-analysis">Active-subnetwork-oriented Enrichment Analysis</a> of this vignette.</p>
<p>The enrichment analysis usually yields a great number of enriched terms whose biological functions are related. Therefore, we implemented two clustering approaches using a pairwise distance matrix based on the kappa statistics between the enriched terms (as proposed by Huang et al. <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>). Based on this distance metric, the user can perform either hierarchical (default) or fuzzy clustering of the enriched terms. Details of clustering and partitioning of enriched terms are presented in the <a href="#clustering-enriched-terms">Clustering Enriched Terms</a> section of this vignette.</p>
<p>Other functionalities of pathfindR including:</p>
<ul>
<li>agglomerated score calculation per each term (to investigate how a gene set is altered in a given sample)</li>
<li>visualization of terms and term-related genes as a graph (to determine the degree of overlap between the enriched terms by identifying shared and/or distinct significant genes) and</li>
<li>pathfindR analysis with custom gene sets are also briefly described.</li>
</ul>
</div>
<div class="section level2">
<h2 id="active-subnetwork-oriented-enrichment-analysis">Active-subnetwork-oriented Enrichment Analysis<a class="anchor" aria-label="anchor" href="#active-subnetwork-oriented-enrichment-analysis"></a>
</h2>
<p><img src="pathfindr.png" style="max-width:100%;"></p>
<p>For convenience, we provide the wrapper function <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> to be used for the active-subnetwork-oriented enrichment analysis. The input for this function must be a data frame consisting of the columns containing: <code>Gene Symbols</code>, <code>Change Values</code> (optional) and <code>p values</code>. The example data frame used in this vignette (<code>input_df</code>) is the dataset containing the differentially-expressed genes for the GEO dataset GSE15573 comparing 18 rheumatoid arthritis (RA) patients versus 15 healthy subjects.</p>
<p>The first 6 rows of the example input data frame are displayed below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://egeulgen.github.io/pathfindR/">pathfindR</a></span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RA_input</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Gene.symbol</th>
<th align="right">logFC</th>
<th align="right">adj.P.Val</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">FAM110A</td>
<td align="right">-0.6939359</td>
<td align="right">0.0000034</td>
</tr>
<tr class="even">
<td align="left">RNASE2</td>
<td align="right">1.3535040</td>
<td align="right">0.0000101</td>
</tr>
<tr class="odd">
<td align="left">S100A8</td>
<td align="right">1.5448338</td>
<td align="right">0.0000347</td>
</tr>
<tr class="even">
<td align="left">S100A9</td>
<td align="right">1.0280904</td>
<td align="right">0.0002263</td>
</tr>
<tr class="odd">
<td align="left">TEX261</td>
<td align="right">-0.3235994</td>
<td align="right">0.0002263</td>
</tr>
<tr class="even">
<td align="left">ARHGAP17</td>
<td align="right">-0.6919330</td>
<td align="right">0.0002708</td>
</tr>
</tbody>
</table>
<blockquote>
<p>For a detailed step-by-step explanation and an unwrapped demonstration of the active-subnetwork-oriented enrichment analysis, see the vignette <a href="manual_execution.html">Step-by-Step Execution of the pathfindR Enrichment Workflow</a></p>
</blockquote>
<p>Executing the workflow is straightforward (but does typically take several minutes):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="useful-arguments">Useful arguments<a class="anchor" aria-label="anchor" href="#useful-arguments"></a>
</h3>
<p>This subsection demonstrates some (selected) useful arguments of <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code>. For a full list of arguments, see <code><a href="../reference/run_pathfindr.html">?run_pathfindR</a></code> or visit <a href="https://github.com/egeulgen/pathfindR/wiki" class="external-link">our GitHub wiki</a>.</p>
<div class="section level4">
<h4 id="filtering-input-genes">Filtering Input Genes<a class="anchor" aria-label="anchor" href="#filtering-input-genes"></a>
</h4>
<p>By default, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> uses the input genes with p-values &lt; 0.05. To change this threshold, use <code>p_val_threshold</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, p_val_threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="output-directory">Output Directory<a class="anchor" aria-label="anchor" href="#output-directory"></a>
</h4>
<p>By default, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> creates a directory named <code>"pathfindR_Results"</code> under the current working directory for writing the output files. To change the output directory, use <code>output_dir</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, output_dir <span class="op">=</span> <span class="st">"this_is_my_output_directory"</span><span class="op">)</span></span></code></pre></div>
<p>This creates <code>"this_is_my_output_directory"</code> under the current working directory.</p>
<p>In essence, this argument is treated as a path so it can be used to create the output directory anywhere. For example, to create the directory <code>"my_dir"</code> under <code>"~/Desktop"</code> and run the analysis there, you may run:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, output_dir <span class="op">=</span> <span class="st">"~/Desktop/my_dir"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Note: If the output directory (e.g. <code>"my_dir"</code>) already exists, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> creates and works under <code>"my_dir(1)"</code>. If that exists also exists, it creates <code>"my_dir(2)"</code> and so on. This was intentionally implemented so that any previous pathfindR results are not overwritten.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="gene-sets-for-enrichment">Gene Sets for Enrichment<a class="anchor" aria-label="anchor" href="#gene-sets-for-enrichment"></a>
</h4>
<p>The active-subnetwork-oriented enrichment analyses can be performed on any gene sets (biological pathways, gene ontology terms, transcription factor target genes, miRNA target genes etc.). The available gene sets in pathfindR are “KEGG”, “Reactome”, “BioCarta”, “GO-All”, “GO-BP”, “GO-CC” and “GO-MF” (all for Homo sapiens). For changing the default gene sets for enrichment analysis (hsa KEGG pathways), use the argument <code>gene_sets</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, gene_sets <span class="op">=</span> <span class="st">"GO-MF"</span><span class="op">)</span></span></code></pre></div>
<p>By default, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> filters the gene sets by including only the terms containing at least 10 and at most 300 genes. To change the default behavior, you may change <code>min_gset_size</code> and <code>max_gset_size</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Including more terms for enrichment analysis</span></span>
<span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, </span>
<span>                           gene_sets <span class="op">=</span> <span class="st">"GO-MF"</span>,</span>
<span>                           min_gset_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                           max_gset_size <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Note that increasing the number of terms for enrichment analysis may result in significantly longer run time.</p>
</blockquote>
<p>If the user prefers to use another gene set source, the <code>gene_sets</code> argument should be set to <code>"Custom"</code> and the custom gene sets (list) and the custom gene set descriptions (named vector) should be supplied via the arguments <code>custom_genes</code> and <code>custom_descriptions</code>, respectively. See <code><a href="../reference/fetch_gene_set.html">?fetch_gene_set</a></code> for more details and <a href="#analysis-with-custom-gene-sets">Analysis with Custom Gene Sets</a> for a simple demonstration.</p>
<p>For details on obtaining organism-specific Gene Sets and PIN data, see the vignette <a href="obtain_data.html">Obtaining PIN and Gene Sets Data</a>.</p>
</div>
<div class="section level4">
<h4 id="filtering-enriched-terms-by-adjusted-p-values">Filtering Enriched Terms by Adjusted-p Values<a class="anchor" aria-label="anchor" href="#filtering-enriched-terms-by-adjusted-p-values"></a>
</h4>
<p>By default, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> adjusts the enrichment p values via the “bonferroni” method and filters the enriched terms by adjusted-p value &lt; 0.05. To change this adjustment method and the threshold, set <code>adj_method</code> and <code>enrichment_threshold</code>, respectively:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, </span>
<span>                           adj_method <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>                           enrichment_threshold <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="protein-protein-interaction-network">Protein-protein Interaction Network<a class="anchor" aria-label="anchor" href="#protein-protein-interaction-network"></a>
</h4>
<p>For the active subnetwork search process, a protein-protein interaction network (PIN) is used. <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> maps the input genes onto this PIN and identifies active subnetworks which are then be used for enrichment analyses. To change the default PIN (“Biogrid”), use the <code>pin_name_path</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, pin_name_path <span class="op">=</span> <span class="st">"IntAct"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>pin_name_path</code> argument can be one of “Biogrid”, “STRING”, “GeneMania”, “IntAct”, “KEGG”, “mmu_STRING” or it can be the path to a custom PIN file provided by the user.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># to use an external PIN of your choice</span></span>
<span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, pin_name_path <span class="op">=</span> <span class="st">"/path/to/myPIN.sif"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>NOTE: the PIN is also used for generating the background genes (in this case, all unique genes in the PIN) during hypergeometric-distribution-based tests in enrichment analyses. Therefore, a large PIN will generally result in better results.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="active-subnetwork-search-method">Active Subnetwork Search Method<a class="anchor" aria-label="anchor" href="#active-subnetwork-search-method"></a>
</h4>
<p>Currently, there are three algorithms implemented in pathfindR for active subnetwork search: Greedy Algorithm (default, based on Ideker et al. <a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>), Simulated Annealing Algorithm (based on Ideker et al. <a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>) and Genetic Algorithm (based on Ozisik et al. <a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>). For a detailed discussion on which algorithm to use see <a href="https://github.com/egeulgen/pathfindR/wiki/Active-subnetwork-oriented-Enrichment-Documentation#selecting-the-active-subnetwork-search-algorithm" class="external-link">this wiki entry</a></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for simulated annealing:</span></span>
<span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, search_method <span class="op">=</span> <span class="st">"SA"</span><span class="op">)</span></span>
<span><span class="co"># for genetic algorithm:</span></span>
<span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, search_method <span class="op">=</span> <span class="st">"GA"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="other-arguments">Other Arguments<a class="anchor" aria-label="anchor" href="#other-arguments"></a>
</h4>
<p>Because the active subnetwork search algorithms are stochastic, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> may be set to iterate the active subnetwork identification and enrichment steps multiple times (by default 1 time). To change this number, set <code>iterations</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, iterations <span class="op">=</span> <span class="fl">25</span><span class="op">)</span> </span></code></pre></div>
<p><code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> uses a parallel loop (using the package <code>foreach</code>) for performing these iterations in parallel. By default, the number of processes to be used is determined automatically. To override, change <code>n_processes</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if not set, n_processes defaults to (number of detected cores - 1)</span></span>
<span><span class="va">output_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>, iterations <span class="op">=</span> <span class="fl">5</span>, n_processes <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<div class="section level4">
<h4 id="enriched-terms-data-frame">Enriched Terms Data Frame<a class="anchor" aria-label="anchor" href="#enriched-terms-data-frame"></a>
</h4>
<p><code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> returns a data frame of enriched terms. Columns are:</p>
<ul>
<li>ID: ID of the enriched term</li>
<li>Term_Description: Description of the enriched term</li>
<li>Fold_Enrichment: Fold enrichment value for the enriched term (Calculated using ONLY the input genes)</li>
<li>occurrence: The number of iterations that the given term was found to enriched over all iterations</li>
<li>lowest_p: the lowest adjusted-p value of the given term over all iterations</li>
<li>highest_p: the highest adjusted-p value of the given term over all iterations</li>
<li>non_Signif_Snw_Genes (OPTIONAL): the non-significant active subnetwork genes, comma-separated (controlled by <code>list_active_snw_genes</code>, default is <code>FALSE</code>)</li>
<li>Up_regulated: the up-regulated genes (as determined by <code>change value</code> &gt; 0, if the <code>change column</code> was provided) in the input involved in the given term’s gene set, comma-separated. If change column was not provided, all affected input genes are listed here.</li>
<li>Down_regulated: the down-regulated genes (as determined by <code>change value</code> &lt; 0, if the <code>change column</code> was provided) in the input involved in the given term’s gene set, comma-separated</li>
</ul>
<p>The first 2 rows of the output data frame of the example analysis on the rheumatoid arthritis gene-level differential expression input data (<code>RA_input</code>) is shown below:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RA_output</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="4%">
<col width="19%">
<col width="7%">
<col width="5%">
<col width="4%">
<col width="4%">
<col width="4%">
<col width="29%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th align="left">ID</th>
<th align="left">Term_Description</th>
<th align="right">Fold_Enrichment</th>
<th align="right">occurrence</th>
<th align="right">support</th>
<th align="right">lowest_p</th>
<th align="right">highest_p</th>
<th align="left">Up_regulated</th>
<th align="left">Down_regulated</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">hsa05415</td>
<td align="left">Diabetic cardiomyopathy</td>
<td align="right">3.246357</td>
<td align="right">10</td>
<td align="right">0.0752907</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="left">NCF4, MMP9, NDUFA1, NDUFB3, UQCRQ, COX6A1, COX7A2, COX7C, GAPDH</td>
<td align="left">ATP2A2, MTOR, PDHA1, PDHB, VDAC1, SLC25A5, PARP1</td>
</tr>
<tr class="even">
<td align="left">hsa04130</td>
<td align="left">SNARE interactions in vesicular transport</td>
<td align="right">4.599006</td>
<td align="right">10</td>
<td align="right">0.0118360</td>
<td align="right">1e-07</td>
<td align="right">1e-07</td>
<td align="left">STX6</td>
<td align="left">STX2, BET1L, SNAP23</td>
</tr>
</tbody>
</table>
<p>By default, <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> also produces a graphical summary of enrichment results for top 10 enriched terms, which can also be later produced by <code><a href="../reference/enrichment_chart.html">enrichment_chart()</a></code>:</p>
<p><img src="enrichment_chart.png" style="max-width:100%;"></p>
<p>You may also disable plotting this chart by setting <code>plot_enrichment_chart=FALSE</code> and later produce this plot via the function <code><a href="../reference/enrichment_chart.html">enrichment_chart()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># change number of top terms plotted (default = 10)</span></span>
<span><span class="fu"><a href="../reference/enrichment_chart.html">enrichment_chart</a></span><span class="op">(</span>result_df <span class="op">=</span> <span class="va">RA_output</span>, </span>
<span>                 top_terms <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="html-report">HTML Report<a class="anchor" aria-label="anchor" href="#html-report"></a>
</h4>
<p>The function also creates an HTML report <code>results.html</code> that is saved in the output directory. This report contains links to two other HTML files:</p>
<p><strong>1. <code>enriched_terms.html</code></strong></p>
<p>This document contains the table of the active subnetwork-oriented enrichment results (same as the returned data frame). By default, each enriched term description is linked to the visualization of the term, with the gene nodes colored according to their change values. If you choose not to create the visualization files, set <code>visualize_enriched_terms = FALSE</code>.</p>
<p><strong>2. <code>conversion_table.html</code></strong></p>
<p>This document contains the table of converted gene symbols. Columns are:</p>
<ul>
<li>Old Symbol: the original gene symbol</li>
<li>Converted Symbol: the alias symbol that was found in the PIN</li>
<li>Change: the provided change value</li>
<li>p-value: the provided adjusted p value</li>
</ul>
<blockquote>
<p>During input processing, gene symbols that are not in the PIN are identified and excluded. For human genes, if aliases of these missing gene symbols are found in the PIN, these symbols are converted to the corresponding aliases (controlled by the argument <code>convert2alias</code>). This step is performed to best map the input data onto the PIN.</p>
</blockquote>
<p>The document contains a second table of genes for which <strong>no interactions</strong> were identified after checking for alias symbols (so these could not be used during the analysis).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="clustering-enriched-terms">Clustering Enriched Terms<a class="anchor" aria-label="anchor" href="#clustering-enriched-terms"></a>
</h2>
<p><img src="term_clustering.png" style="max-width:100%;"></p>
<p>The wrapper function <code><a href="../reference/cluster_enriched_terms.html">cluster_enriched_terms()</a></code> can be used to perform clustering of enriched terms and partitioning the terms into biologically-relevant groups. Clustering can be performed either via <code>hierarchical</code> or <code>fuzzy</code> method using the pairwise kappa statistics (a chance-corrected measure of co-occurrence between two sets of categorized data) matrix between all enriched terms.</p>
<div class="section level3">
<h3 id="hierarchical-clustering">Hierarchical Clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a>
</h3>
<p>By default, <code><a href="../reference/cluster_enriched_terms.html">cluster_enriched_terms()</a></code> performs hierarchical clustering of the terms (using <span class="math inline">\(1 - \kappa\)</span> as the distance metric). Iterating over <span class="math inline">\(2,3,...n\)</span> clusters (where <span class="math inline">\(n\)</span> is the number of terms), <code><a href="../reference/cluster_enriched_terms.html">cluster_enriched_terms()</a></code> determines the optimal number of clusters by maximizing the average silhouette width, partitions the data into this optimal number of clusters and returns a data frame with cluster assignments.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RA_clustered</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_enriched_terms.html">cluster_enriched_terms</a></span><span class="op">(</span><span class="va">RA_output</span>, plot_dend <span class="op">=</span> <span class="cn">FALSE</span>, plot_clusters_graph <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## First 2 rows of clustered data frame</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">RA_clustered</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="3%">
<col width="11%">
<col width="6%">
<col width="4%">
<col width="4%">
<col width="3%">
<col width="4%">
<col width="27%">
<col width="21%">
<col width="3%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">ID</th>
<th align="left">Term_Description</th>
<th align="right">Fold_Enrichment</th>
<th align="right">occurrence</th>
<th align="right">support</th>
<th align="right">lowest_p</th>
<th align="right">highest_p</th>
<th align="left">Up_regulated</th>
<th align="left">Down_regulated</th>
<th align="right">Cluster</th>
<th align="left">Status</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">hsa05415</td>
<td align="left">Diabetic cardiomyopathy</td>
<td align="right">3.246357</td>
<td align="right">10</td>
<td align="right">0.0752907</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="left">NCF4, MMP9, NDUFA1, NDUFB3, UQCRQ, COX6A1, COX7A2, COX7C, GAPDH</td>
<td align="left">ATP2A2, MTOR, PDHA1, PDHB, VDAC1, SLC25A5, PARP1</td>
<td align="right">1</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">8</td>
<td align="left">hsa00190</td>
<td align="left">Oxidative phosphorylation</td>
<td align="right">2.943760</td>
<td align="right">10</td>
<td align="right">0.0177005</td>
<td align="right">3e-07</td>
<td align="right">3e-07</td>
<td align="left">NDUFA1, NDUFB3, UQCRQ, COX6A1, COX7A2, COX7C, ATP6V1D, ATP6V0E1</td>
<td align="left">ATP6V0E2</td>
<td align="right">1</td>
<td align="left">Member</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## The representative terms</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">RA_clustered</span><span class="op">[</span><span class="va">RA_clustered</span><span class="op">$</span><span class="va">Status</span> <span class="op">==</span> <span class="st">"Representative"</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="3%">
<col width="14%">
<col width="5%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="3%">
<col width="22%">
<col width="29%">
<col width="2%">
<col width="5%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">ID</th>
<th align="left">Term_Description</th>
<th align="right">Fold_Enrichment</th>
<th align="right">occurrence</th>
<th align="right">support</th>
<th align="right">lowest_p</th>
<th align="right">highest_p</th>
<th align="left">Up_regulated</th>
<th align="left">Down_regulated</th>
<th align="right">Cluster</th>
<th align="left">Status</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">hsa05415</td>
<td align="left">Diabetic cardiomyopathy</td>
<td align="right">3.2463571</td>
<td align="right">10</td>
<td align="right">0.0752907</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="left">NCF4, MMP9, NDUFA1, NDUFB3, UQCRQ, COX6A1, COX7A2, COX7C, GAPDH</td>
<td align="left">ATP2A2, MTOR, PDHA1, PDHB, VDAC1, SLC25A5, PARP1</td>
<td align="right">1</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">hsa04130</td>
<td align="left">SNARE interactions in vesicular transport</td>
<td align="right">4.5990059</td>
<td align="right">10</td>
<td align="right">0.0118360</td>
<td align="right">0.0000001</td>
<td align="right">0.0000001</td>
<td align="left">STX6</td>
<td align="left">STX2, BET1L, SNAP23</td>
<td align="right">2</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">hsa03410</td>
<td align="left">Base excision repair</td>
<td align="right">5.7487574</td>
<td align="right">1</td>
<td align="right">0.0053191</td>
<td align="right">0.0000001</td>
<td align="right">0.0000001</td>
<td align="left">POLE4</td>
<td align="left">MUTYH, APEX2, POLD2, PARP1</td>
<td align="right">3</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">4</td>
<td align="left">hsa04064</td>
<td align="left">NF-kappa B signaling pathway</td>
<td align="right">2.9185999</td>
<td align="right">10</td>
<td align="right">0.0502592</td>
<td align="right">0.0000001</td>
<td align="right">0.0000001</td>
<td align="left">LY96</td>
<td align="left">PRKCQ, CARD11, TICAM1, IKBKB, PARP1, UBE2I, CSNK2A2</td>
<td align="right">4</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">5</td>
<td align="left">hsa03040</td>
<td align="left">Spliceosome</td>
<td align="right">3.6887860</td>
<td align="right">10</td>
<td align="right">0.0477457</td>
<td align="right">0.0000002</td>
<td align="right">0.0000446</td>
<td align="left">SF3B6, LSM3, BUD31</td>
<td align="left">SNRPB, SF3B2, U2AF2, PUF60, SNU13, DDX23, EIF4A3, HNRNPA1, PCBP1, SRSF8, SRSF5</td>
<td align="right">5</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">6</td>
<td align="left">hsa04714</td>
<td align="left">Thermogenesis</td>
<td align="right">2.5174653</td>
<td align="right">10</td>
<td align="right">0.0441982</td>
<td align="right">0.0000003</td>
<td align="right">0.0000003</td>
<td align="left">NDUFA1, NDUFB3, UQCRQ, COX6A1, COX7A2, COX7C</td>
<td align="left">ADCY7, CREB1, KDM1A, SMARCA4, ACTG1, ACTB, ARID1A, MTOR</td>
<td align="right">6</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">7</td>
<td align="left">hsa03013</td>
<td align="left">RNA transport</td>
<td align="right">2.6017234</td>
<td align="right">10</td>
<td align="right">0.0184111</td>
<td align="right">0.0000003</td>
<td align="right">0.0000419</td>
<td align="left">NUP214</td>
<td align="left">NUP62, NUP93, RANGAP1, UBE2I, SUMO3, GEMIN4, EIF2S3, EIF2B1, EIF4A3, RNPS1, SRRM1</td>
<td align="right">7</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">15</td>
<td align="left">hsa04722</td>
<td align="left">Neurotrophin signaling pathway</td>
<td align="right">2.8938660</td>
<td align="right">10</td>
<td align="right">0.0414102</td>
<td align="right">0.0000088</td>
<td align="right">0.0000151</td>
<td align="left"></td>
<td align="left">SH2B3, CRKL, FASLG, CALM3, CALM1, ABL1, MAGED1, IRAK2, IKBKB</td>
<td align="right">8</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">17</td>
<td align="left">hsa04630</td>
<td align="left">JAK-STAT signaling pathway</td>
<td align="right">1.4500051</td>
<td align="right">10</td>
<td align="right">0.0115943</td>
<td align="right">0.0000171</td>
<td align="right">0.0000171</td>
<td align="left"></td>
<td align="left">IL2RB, IL10RA, IL27RA, JAK1, PIAS3, MTOR</td>
<td align="right">9</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">23</td>
<td align="left">hsa05203</td>
<td align="left">Viral carcinogenesis</td>
<td align="right">1.8396024</td>
<td align="right">10</td>
<td align="right">0.0346832</td>
<td align="right">0.0000446</td>
<td align="right">0.0000446</td>
<td align="left">GTF2B</td>
<td align="left">CREB1, JAK1, SCRIB, RBL2, HDAC1, DNAJA3, SRF</td>
<td align="right">10</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">33</td>
<td align="left">hsa04110</td>
<td align="left">Cell cycle</td>
<td align="right">1.5299112</td>
<td align="right">10</td>
<td align="right">0.0118003</td>
<td align="right">0.0002246</td>
<td align="right">0.0004249</td>
<td align="left"></td>
<td align="left">RBL2, ABL1, HDAC1, CDKN1C, ANAPC1</td>
<td align="right">11</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">41</td>
<td align="left">hsa03010</td>
<td align="left">Ribosome</td>
<td align="right">1.7473197</td>
<td align="right">9</td>
<td align="right">0.0058480</td>
<td align="right">0.0007058</td>
<td align="right">0.0459772</td>
<td align="left">MRPS18C, RPS24, MRPL33, RPL26, RPL31, RPL39</td>
<td align="left">RPLP2</td>
<td align="right">12</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">42</td>
<td align="left">hsa00020</td>
<td align="left">Citrate cycle (TCA cycle)</td>
<td align="right">3.7941799</td>
<td align="right">10</td>
<td align="right">0.0179106</td>
<td align="right">0.0007264</td>
<td align="right">0.0009951</td>
<td align="left"></td>
<td align="left">MDH2, PDHA1, PDHB</td>
<td align="right">13</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">48</td>
<td align="left">hsa00630</td>
<td align="left">Glyoxylate and dicarboxylate metabolism</td>
<td align="right">2.6166758</td>
<td align="right">2</td>
<td align="right">0.0060061</td>
<td align="right">0.0011912</td>
<td align="right">0.0015442</td>
<td align="left"></td>
<td align="left">MDH2, SHMT1</td>
<td align="right">14</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">52</td>
<td align="left">hsa03050</td>
<td align="left">Proteasome</td>
<td align="right">1.6863022</td>
<td align="right">10</td>
<td align="right">0.0119404</td>
<td align="right">0.0014614</td>
<td align="right">0.0014614</td>
<td align="left"></td>
<td align="left">PSMD7, PSMB10</td>
<td align="right">15</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">64</td>
<td align="left">hsa05202</td>
<td align="left">Transcriptional misregulation in cancer</td>
<td align="right">2.2188187</td>
<td align="right">10</td>
<td align="right">0.0058310</td>
<td align="right">0.0033621</td>
<td align="right">0.0033621</td>
<td align="left">MMP9, DDIT3</td>
<td align="left">HDAC1, SIN3A, BCL11B, SLC45A3, EWSR1, IL2RB, TAF15, ASPSCR1</td>
<td align="right">16</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">96</td>
<td align="left">hsa04120</td>
<td align="left">Ubiquitin mediated proteolysis</td>
<td align="right">1.6616846</td>
<td align="right">10</td>
<td align="right">0.0058310</td>
<td align="right">0.0214469</td>
<td align="right">0.0414098</td>
<td align="left">TRIP12</td>
<td align="left">UBE2G1, UBE2I, HERC1, PIAS3, ANAPC1</td>
<td align="right">17</td>
<td align="left">Representative</td>
</tr>
<tr class="even">
<td align="left">101</td>
<td align="left">hsa00340</td>
<td align="left">Histidine metabolism</td>
<td align="right">5.4202570</td>
<td align="right">10</td>
<td align="right">0.0058310</td>
<td align="right">0.0335915</td>
<td align="right">0.0335915</td>
<td align="left">HNMT</td>
<td align="left">ALDH9A1, CNDP2</td>
<td align="right">18</td>
<td align="left">Representative</td>
</tr>
<tr class="odd">
<td align="left">108</td>
<td align="left">hsa04350</td>
<td align="left">TGF-beta signaling pathway</td>
<td align="right">0.8526247</td>
<td align="right">9</td>
<td align="right">0.0058140</td>
<td align="right">0.0404554</td>
<td align="right">0.0458347</td>
<td align="left"></td>
<td align="left">SMAD7, TGIF2</td>
<td align="right">19</td>
<td align="left">Representative</td>
</tr>
</tbody>
</table>
<p>After clustering, you may again plot the summary enrichment chart and display the enriched terms by clusters:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plotting only selected clusters for better visualization</span></span>
<span><span class="va">RA_selected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">RA_clustered</span>, <span class="va">Cluster</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">5</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/enrichment_chart.html">enrichment_chart</a></span><span class="op">(</span><span class="va">RA_selected</span>, plot_by_cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_vignette_files/figure-html/hierarchical2-1.png" width="672" style="display: block; margin: auto;"></p>
<p>For details, see <code><a href="../reference/hierarchical_term_clustering.html">?hierarchical_term_clustering</a></code></p>
</div>
<div class="section level3">
<h3 id="heuristic-fuzzy-multiple-linkage-partitioning">Heuristic Fuzzy Multiple-linkage Partitioning<a class="anchor" aria-label="anchor" href="#heuristic-fuzzy-multiple-linkage-partitioning"></a>
</h3>
<p>Alternatively, the <code>fuzzy</code> clustering method (as described by Huang et al.<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>) can be used:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RA_clustered_fuzzy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cluster_enriched_terms.html">cluster_enriched_terms</a></span><span class="op">(</span><span class="va">RA_output</span>, method <span class="op">=</span> <span class="st">"fuzzy"</span><span class="op">)</span></span></code></pre></div>
<p>For details, see <code><a href="../reference/fuzzy_term_clustering.html">?fuzzy_term_clustering</a></code></p>
</div>
</div>
<div class="section level2">
<h2 id="term-gene-heatmap">Term-Gene Heatmap<a class="anchor" aria-label="anchor" href="#term-gene-heatmap"></a>
</h2>
<p>The function <code><a href="../reference/term_gene_heatmap.html">term_gene_heatmap()</a></code> can be used to visualize the heatmap of genes that are involved in the enriched terms. This heatmap allows visual identification of the input genes involved in the enriched terms, as well as the common or distinct genes between different terms. If the input data frame (same as in <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code>) is supplied, the tile colors indicate the change values.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/term_gene_heatmap.html">term_gene_heatmap</a></span><span class="op">(</span>result_df <span class="op">=</span> <span class="va">RA_output</span>, genes_df <span class="op">=</span> <span class="va">RA_input</span><span class="op">)</span></span></code></pre></div>
<p><img src="hmap.png" style="max-width:100%;"></p>
<p>See the vignette <a href="visualization_vignette.html">Visualization of pathfindR Enrichment Results</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="term-gene-graph">Term-Gene Graph<a class="anchor" aria-label="anchor" href="#term-gene-graph"></a>
</h2>
<p>The visualization function <code><a href="../reference/term_gene_graph.html">term_gene_graph()</a></code> (adapted from the “Gene-Concept network visualization” by the R package <code>enrichplot</code>) can be utilized to visualize which genes are involved in the enriched terms. The function creates a term-gene graph which shows the connections between genes and biological terms (enriched pathways or gene sets). This allows for the investigation of multiple terms to which significant genes are related. This graph also enables visual determination of the degree of overlap between the enriched terms by identifying shared and/or distinct significant genes.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/term_gene_graph.html">term_gene_graph</a></span><span class="op">(</span>result_df <span class="op">=</span> <span class="va">RA_output</span>, use_description <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="term_gene.png" style="max-width:100%;"></p>
<p>See the vignette <a href="visualization_vignette.html">Visualization of pathfindR Enrichment Results</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="upset-plot">UpSet Plot<a class="anchor" aria-label="anchor" href="#upset-plot"></a>
</h2>
<p>UpSet plots are plots of the intersections of sets as a matrix. This function creates a ggplot object of an UpSet plot where the x-axis is the UpSet plot of intersections of enriched terms. By default (i.e., <code>method = "heatmap"</code>), the main plot is a heatmap of genes at the corresponding intersections, colored by up/down regulation (if <code>genes_df</code> is provided, colored by change values). If <code>method = "barplot"</code>, the main plot is bar plots of the number of genes at the corresponding intersections. Finally, if <code>method = "boxplot"</code> and <code>genes_df</code> is provided, then the main plot displays the boxplots of change values of the genes at the corresponding intersections.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/UpSet_plot.html">UpSet_plot</a></span><span class="op">(</span>result_df <span class="op">=</span> <span class="va">RA_output</span>, genes_df <span class="op">=</span> <span class="va">RA_input</span><span class="op">)</span></span></code></pre></div>
<p><img src="upset.png" style="max-width:100%;"></p>
<p>See the vignette <a href="visualization_vignette.html">Visualization of pathfindR Enrichment Results</a> for more details.</p>
</div>
<div class="section level2">
<h2 id="aggregated-term-scores-per-sample">Aggregated Term Scores per Sample<a class="anchor" aria-label="anchor" href="#aggregated-term-scores-per-sample"></a>
</h2>
<p>The function <code><a href="../reference/score_terms.html">score_terms()</a></code> can be used to calculate the agglomerated z score of each enriched term per sample. This allows the user to individually examine the scores and infer how a term is overall altered (activated or repressed) in a given sample or a group of samples.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Vector of "Case" IDs</span></span>
<span><span class="va">cases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GSM389703"</span>, <span class="st">"GSM389704"</span>, <span class="st">"GSM389706"</span>, <span class="st">"GSM389708"</span>, </span>
<span>           <span class="st">"GSM389711"</span>, <span class="st">"GSM389714"</span>, <span class="st">"GSM389716"</span>, <span class="st">"GSM389717"</span>, </span>
<span>           <span class="st">"GSM389719"</span>, <span class="st">"GSM389721"</span>, <span class="st">"GSM389722"</span>, <span class="st">"GSM389724"</span>, </span>
<span>           <span class="st">"GSM389726"</span>, <span class="st">"GSM389727"</span>, <span class="st">"GSM389730"</span>, <span class="st">"GSM389731"</span>, </span>
<span>           <span class="st">"GSM389733"</span>, <span class="st">"GSM389735"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate scores for representative terms </span></span>
<span><span class="co">## and plot heat map using term descriptions</span></span>
<span><span class="va">score_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score_terms.html">score_terms</a></span><span class="op">(</span>enrichment_table <span class="op">=</span> <span class="va">RA_clustered</span><span class="op">[</span><span class="va">RA_clustered</span><span class="op">$</span><span class="va">Status</span> <span class="op">==</span> <span class="st">"Representative"</span>, <span class="op">]</span>,</span>
<span>                            exp_mat <span class="op">=</span> <span class="va">RA_exp_mat</span>,</span>
<span>                            cases <span class="op">=</span> <span class="va">cases</span>,</span>
<span>                            use_description <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co"># default FALSE</span></span>
<span>                            label_samples <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># default = TRUE</span></span>
<span>                            case_title <span class="op">=</span> <span class="st">"RA"</span>,  <span class="co"># default = "Case"</span></span>
<span>                            control_title <span class="op">=</span> <span class="st">"Healthy"</span>, <span class="co"># default = "Control"</span></span>
<span>                            low <span class="op">=</span> <span class="st">"#f7797d"</span>, <span class="co"># default = "green"</span></span>
<span>                            mid <span class="op">=</span> <span class="st">"#fffde4"</span>, <span class="co"># default = "black"</span></span>
<span>                            high <span class="op">=</span> <span class="st">"#1f4037"</span><span class="op">)</span> <span class="co"># default = "red"</span></span></code></pre></div>
<p><img src="intro_vignette_files/figure-html/scores-1.png" width="768" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="comparison-of-2-pathfindr-results">Comparison of 2 pathfindR Results<a class="anchor" aria-label="anchor" href="#comparison-of-2-pathfindr-results"></a>
</h2>
<p>The function <code><a href="../reference/combine_pathfindR_results.html">combine_pathfindR_results()</a></code> allows combination of two pathfindR active-subnetwork-oriented enrichment analysis results for investigating common and distinct terms between the groups. Below is an example for comparing results using two different rheumatoid arthritis-related data sets(<code>RA_output</code> and <code>RA_comparison_output</code>).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_pathfindR_results.html">combine_pathfindR_results</a></span><span class="op">(</span>result_A <span class="op">=</span> <span class="va">RA_output</span>, </span>
<span>                                         result_B <span class="op">=</span> <span class="va">RA_comparison_output</span>, </span>
<span>                                         plot_common <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; You may run `combined_results_graph()` to create visualizations of combined term-gene graphs of selected terms</span></span></code></pre></div>
<p>For more details, see the vignette <a href="comparing_results.html">Comparing Two pathfindR Results</a></p>
</div>
<div class="section level2">
<h2 id="analysis-with-custom-gene-sets">Analysis with Custom Gene Sets<a class="anchor" aria-label="anchor" href="#analysis-with-custom-gene-sets"></a>
</h2>
<blockquote>
<p>As of v1.5, pathfindR offers utility functions for obtaining organism-specific PIN data and organism-specific gene sets data via <code><a href="../reference/get_pin_file.html">get_pin_file()</a></code> and <code><a href="../reference/get_gene_sets_list.html">get_gene_sets_list()</a></code>, respectively. See the vignette <a href="obtain_data.html">Obtaining PIN and Gene Sets Data</a> for detailed information on how to gather PIN and gene sets data (for any organism of your choice) for use with pathfindR.</p>
</blockquote>
<p>It is possible to use <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> with custom gene sets (including gene sets for non-Homo-sapiens species). Here, we provide an example application of active-subnetwork-oriented enrichment analysis of the target genes of two transcription factors.</p>
<p>We first load and prepare the gene sets:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## CREB target genes</span></span>
<span><span class="va">CREB_target_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/normalizePath.html" class="external-link">normalizePath</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/CREB.txt"</span>, package <span class="op">=</span> <span class="st">"pathfindR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">CREB_target_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">CREB_target_genes</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="co"># skip the first two lines</span></span>
<span></span>
<span><span class="co">## MYC target genes</span></span>
<span><span class="va">MYC_target_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/normalizePath.html" class="external-link">normalizePath</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/MYC.txt"</span>, package <span class="op">=</span> <span class="st">"pathfindR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">MYC_target_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">MYC_target_genes</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="co"># skip the first two lines</span></span>
<span></span>
<span><span class="co">## Prep for use</span></span>
<span><span class="va">custom_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="va">CREB_target_genes</span>, TF2 <span class="op">=</span> <span class="va">MYC_target_genes</span><span class="op">)</span></span>
<span><span class="va">custom_descriptions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>TF1 <span class="op">=</span> <span class="st">"CREB target genes"</span>, TF2 <span class="op">=</span> <span class="st">"MYC target genes"</span><span class="op">)</span></span></code></pre></div>
<p>We next prepare the example input data frame. Because of the way we choose genes, we expect significant enrichment for MYC targets (40 MYC target genes + 10 CREB target genes). Because this is only an example, we also assign each genes random p-values between 0.001 and 0.05.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Select 40 random genes from MYC gene sets and 10 from CREB gene sets</span></span>
<span><span class="va">selected_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">MYC_target_genes</span>, <span class="fl">40</span><span class="op">)</span></span>
<span><span class="va">selected_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">selected_genes</span>, </span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">CREB_target_genes</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Assign random p value between 0.001 and 0.05 for each selected gene</span></span>
<span><span class="va">rand_p_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.001</span>, <span class="fl">0.05</span>, length.out <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>                      size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">selected_genes</span><span class="op">)</span>,</span>
<span>                      replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">input_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Gene_symbol <span class="op">=</span> <span class="va">selected_genes</span>,</span>
<span>                       p_val <span class="op">=</span> <span class="va">rand_p_vals</span><span class="op">)</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input_df</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">Gene_symbol</th>
<th align="right">p_val</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">HNRNPD</td>
<td align="right">0.01325</td>
</tr>
<tr class="even">
<td align="left">IL1RAPL1</td>
<td align="right">0.00100</td>
</tr>
<tr class="odd">
<td align="left">CD3EAP</td>
<td align="right">0.00100</td>
</tr>
<tr class="even">
<td align="left">LTBR</td>
<td align="right">0.02550</td>
</tr>
<tr class="odd">
<td align="left">CGREF1</td>
<td align="right">0.00100</td>
</tr>
<tr class="even">
<td align="left">TPM2</td>
<td align="right">0.05000</td>
</tr>
</tbody>
</table>
<p>Finally, we perform active-subnetwork-oriented enrichment analysis via <code><a href="../reference/run_pathfindr.html">run_pathfindR()</a></code> using the custom genes as the gene sets:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">custom_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_pathfindr.html">run_pathfindR</a></span><span class="op">(</span><span class="va">input_df</span>,</span>
<span>                               gene_sets <span class="op">=</span> <span class="st">"Custom"</span>,</span>
<span>                               custom_genes <span class="op">=</span> <span class="va">custom_genes</span>,</span>
<span>                               custom_descriptions <span class="op">=</span> <span class="va">custom_descriptions</span>,</span>
<span>                               max_gset_size <span class="op">=</span> <span class="cn">Inf</span>, <span class="co"># DO NOT LIMIT GENE SET SIZE</span></span>
<span>                               output_dir <span class="op">=</span> <span class="st">"misc/v1.4/CREB_MYC"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">custom_result</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="1%">
<col width="4%">
<col width="4%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="2%">
<col width="75%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left">ID</th>
<th align="left">Term_Description</th>
<th align="right">Fold_Enrichment</th>
<th align="right">occurrence</th>
<th align="right">support</th>
<th align="right">lowest_p</th>
<th align="right">highest_p</th>
<th align="left">Up_regulated</th>
<th align="left">Down_regulated</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">TF2</td>
<td align="left">MYC target genes</td>
<td align="right">18.23368</td>
<td align="right">1</td>
<td align="right">0.8072289</td>
<td align="right">0.00e+00</td>
<td align="right">0.00e+00</td>
<td align="left">AGRP, ATP6V1C1, C19orf54, CD3EAP, CGREF1, CNPY3, EPB41, FOXD3, FXR1, GLA, HNRNPD, HOXA7, IL1RAPL1, KDM6A, LONP1, LTBR, MDN1, MICU1, NET1, NEUROD6, NMNAT2, NOL6, NUDC, PEPD, PKN1, PSMB3, RPS19, RPS28, RRS1, SLC9A5, SMC3, STC2, TESK2, TNPO2, TOPORS, TPM2, TSSK3, WBP2, ZBTB8OS, ZFYVE26, ZHX2</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">TF1</td>
<td align="left">CREB target genes</td>
<td align="right">20.46776</td>
<td align="right">1</td>
<td align="right">0.2530120</td>
<td align="right">1.26e-05</td>
<td align="right">1.26e-05</td>
<td align="left">BRAF, DIO2, ELAVL1, EPB41, FAM65A, FOXD3, NEUROD6, NOC4L, NUPL2, PPP1R15A, SYNGR3, TIPRL</td>
<td align="left"></td>
</tr>
</tbody>
</table>
<blockquote>
<p>It is also possible to run pathfindR using non-human organism annotation. See the vignette <a href="non_hs_analysis.html">pathfindR Analysis for non-Homo-sapiens organisms</a></p>
</blockquote>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Huang DW, Sherman BT, Tan Q, et al. The DAVID Gene Functional Classification Tool: a novel biological module-centric algorithm to functionally analyze large gene lists. Genome Biol. 2007;8(9):R183.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Ideker T, Ozier O, Schwikowski B, Siegel AF. Discovering regulatory and signalling circuits in molecular interaction networks. Bioinformatics. 2002;18 Suppl 1:S233-40.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Ideker T, Ozier O, Schwikowski B, Siegel AF. Discovering regulatory and signalling circuits in molecular interaction networks. Bioinformatics. 2002;18 Suppl 1:S233-40.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>Ozisik O, Bakir-Gungor B, Diri B, Sezerman OU. Active Subnetwork GA: A Two Stage Genetic Algorithm Approach to Active Subnetwork Search. Current Bioinformatics. 2017; 12(4):320-8. 10.2174/1574893611666160527100444<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>Huang DW, Sherman BT, Tan Q, et al. The DAVID Gene Functional Classification Tool: a novel biological module-centric algorithm to functionally analyze large gene lists. Genome Biol. 2007;8(9):R183.<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Ege Ulgen, Ozan Ozisik.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.9000.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '7f13d388d59fea08d4add29291ea2896',
    indexName: 'pathfindr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
